pkg/providers/tinkerbell/comparator.go

```diff
@@ -0,0 +1,30 @@
+package tinkerbell
+
+import (
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+)
+
+// ControlPlaneVersionChange detects kubernetes version has changed between new and old KubeadminControlPlane objects.
+func ControlPlaneVersionChange(oldKCP *controlplanev1.KubeadmControlPlane, newKCP *controlplanev1.KubeadmControlPlane) bool {
+	if oldKCP != nil && newKCP != nil {
```

Comment:
why do we need to check for nil?

Reply:
I'll take the nil check out here. I had the nil checks to allow me to pass values from funcs like `GetKubeadmControlPlane()` directly to the comparison but that is mixing logic.